// @pushprotocol/restapi/src/lib/user/createUser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = void 0;
const tslib_1 = require("tslib");
const helpers_1 = require("../chat/helpers");
const constants_1 = require("../constants");
const helpers_2 = require("../helpers");
const create = (options) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const { env = constants_1.default.ENV.PROD, account = null, signer = null, version = constants_1.default.ENC_TYPE_V3, progressHook, } = options || {};
    try {
        if (account == null && signer == null) {
            throw new Error(`At least one from account or signer is necessary!`);
        }
        const wallet = (0, helpers_1.getWallet)({ account, signer });
        const address = yield (0, helpers_1.getAccountAddress)(wallet);
        if (!(0, helpers_2.isValidETHAddress)(address)) {
            throw new Error(`Invalid address!`);
        }
        const caip10 = (0, helpers_2.walletToPCAIP10)(address);
        let encryptionType = version;
        // falback to v1
        if (!signer)
            encryptionType = constants_1.default.ENC_TYPE_V1;
        // Report Progress
        progressHook === null || progressHook === void 0 ? void 0 : progressHook({
            progressId: 'PUSH-CREATE-01',
            progressTitle: 'Generating Secure Profile Signature',
            progressInfo: 'This step is only done for first time users and might take a few seconds. PGP keys are getting generated to provide you with secure yet seamless chat',
            level: 'INFO',
        });
        const keyPairs = yield (0, helpers_1.generateKeyPair)();
        // Report Progress
        progressHook === null || progressHook === void 0 ? void 0 : progressHook({
            progressId: 'PUSH-CREATE-02',
            progressTitle: 'Signing Generated Profile',
            progressInfo: 'This step is only done for first time users. Please sign the message to continue.',
            level: 'INFO',
        });
        const publicKey = yield (0, helpers_2.preparePGPPublicKey)(encryptionType, keyPairs.publicKeyArmored, wallet);
        // Report Progress
        progressHook === null || progressHook === void 0 ? void 0 : progressHook({
            progressId: 'PUSH-CREATE-03',
            progressTitle: 'Encrypting Generated Profile',
            progressInfo: 'Encrypting your keys. Please sign the message to continue.',
            level: 'INFO',
        });
        const encryptedPrivateKey = yield (0, helpers_2.encryptPGPKey)(encryptionType, keyPairs.privateKeyArmored, address, wallet);
        const body = {
            user: caip10,
            wallet,
            publicKey: publicKey,
            encryptedPrivateKey: JSON.stringify(encryptedPrivateKey),
            encryptionType: encryptionType,
            env,
            encryptedPassword: null,
            nftOwner: null,
        };
        // Report Progress
        progressHook === null || progressHook === void 0 ? void 0 : progressHook({
            progressId: 'PUSH-CREATE-04',
            progressTitle: 'Syncing Generated Profile',
            progressInfo: 'Please sign the message to continue. Steady lads, chat is almost ready!',
            level: 'INFO',
        });
        const createdUser = yield (0, helpers_1.createUserService)(body);
        // Report Progress
        progressHook === null || progressHook === void 0 ? void 0 : progressHook({
            progressId: 'PUSH-CREATE-05',
            progressTitle: 'Setup Complete',
            progressInfo: '',
            level: 'SUCCESS',
        });
        return createdUser;
    }
    catch (err) {
        progressHook === null || progressHook === void 0 ? void 0 : progressHook({
            progressId: 'PUSH-ERROR-00',
            progressTitle: 'Non Specific Error',
            progressInfo: `[Push SDK] - API  - Error - API create User() -: ${err}`,
            level: 'ERROR',
        });
        throw Error(`[Push SDK] - API  - Error - API create User() -: ${err}`);
    }
});
exports.create = create;
